README
NAME
    Perl::Critic::Policy::logicLAB::ProhibitUseLib - simple policy
    prohibiting the use of 'use lib'

AFFILIATION
    This policy is a policy in the Perl::Critic::logicLAB distribution.

VERSION
    This documentation describes version 0.01

DESCRIPTION
    The 'use lib' statement, hardcodes the include path to be used. This can
    give issues when moving modules and scripts between diverse
    environments.

        use lib '/some/path';                                       #not ok
        use lib qw(/you/do/not/want/to/go/down/this/path /or/this); #not ok

    Instead use the environment variable PERL5LIB

        #bash
        export PERL5LIB='/some/path/some/where'
    
        #tcsh and csh
        setenv PERL5LIB '/some/path/some/where'

CONFIGURATION AND ENVIRONMENT
    This Policy is not configurable except for the standard options.

DEPENDENCIES AND REQUIREMENTS
    * Perl::Critic
    * Perl::Critic::Utils
    * Readonly
    * Test::More
    * Test::Perl::Critic

INCOMPATIBILITIES
    This distribution has no known incompatibilities.

BUGS AND LIMITATIONS
    Currently the policy has no special opinion on FindBin. It only aims to
    address messy, misleading, buggy and obscuring use of 'use lib'.

BUG REPORTING
    Please use Requets Tracker for bug reporting:

        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-Prohibit-UseLib

TEST AND QUALITY
    The following policies have been disabled for this distribution

    * Perl::Crititc::Policy::ValuesAndExpressions::ProhibitConstantPragma
    * Perl::Crititc::Policy::NamingConventions::Capitalization

    See also t/perlcriticrc

  TEST COVERAGE
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        File                           stmt   bran   cond    sub    pod   time  total
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        ...ogicLAB/ProhibitUseLib.pm  100.0   50.0    n/a  100.0  100.0  100.0   95.3
        Total                         100.0   50.0    n/a  100.0  100.0  100.0   95.3
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
    
SEE ALSO
    * http://perldoc.perl.org/perlrun.html#ENVIRONMENT
    *
    http://logiclab.jira.com/wiki/display/OPEN/Development#Development-Makey
    ourComponentsEnvironmentAgnostic

AUTHOR
    * Jonas B. Nielsen, jonasbn `<jonasbn@cpan.org>'

LICENSE AND COPYRIGHT
    Copyright (c) 2009-2010 Jonas B. Nielsen. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Changes file for Perl::Critic::Policy::logicLAB::ProhibitUseLib

$Id: Changes 8129 2013-07-27 18:12:34Z jonasbn $

Project change log:
http://logicLAB.jira.com/browse/PCPMPUL#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.02 2013-07-27 

- Update not required

- Fixed up Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [PCPMPUL-10]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec
  
  https://logiclab.jira.com/browse/PCPMPUL-10

- Perl::Critic boiler plate updated to latest version [PCPMPUL-6]

  https://logiclab.jira.com/browse/PCPMPUL-6

- Changes Perl::Critic theme to logicLAB as default [PCPMPUL-7]

  https://logiclab.jira.com/browse/PCPMPUL-7


0.01 2010-09-07

- Initial release


Build.PL
# Build.PL for Perl::Critic::JONASBN

# $Id: Build.PL 7185 2010-09-05 07:32:14Z jonasbn $

use strict;
use Module::Build;

my $build = Module::Build->new(
    dist_author          => 'Jonas B. Nielsen, <jonasbn@cpan.org>',
    module_name          => 'Perl::Critic::Policy::logicLAB::ProhibitUseLib',
    license              => 'artistic',
    recursive_test_files => 0,
    create_makefile_pl   => 'traditional',
    create_readme        => 1,
    dist_abstract        => q[simple policy prohibiting the use of 'use lib'],
    requires => {
        'Perl::Critic'        => '0',
        'Perl::Critic::Utils' => '0',
        'Readonly'            => '0',
        'Test::More'          => '0',
        'Test::Perl::Critic'  => '0',
    },
    meta_merge => {
        resources => {
            homepage => 'http://logiclab.jira.com/wiki/display/PCPMPUL/Home',
            bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-Prohibit-UseLib',
            repository => 'http://logicLAB.jira.com/svn/PCPMPUL',
            ProjectChangelog =>
                'http://logiclab.jira.com/browse/PCPMPUL#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel',
            ProjectSite => 'http://logicLAB.jira.com/browse/PCPMPUL',
            ProjectWiki => 'http://logiclab.jira.com/wiki/display/PCPMPUL/Home',
        }
    },
);

$build->create_build_script();

1;


lib/Perl/Critic/Policy/logicLAB/ProhibitUseLib.pm
package Perl::Critic::Policy::logicLAB::ProhibitUseLib;

# $Id: ProhibitUseLib.pm 8128 2013-07-27 18:08:05Z jonasbn $

use strict;
use warnings;
use base 'Perl::Critic::Policy';
use Perl::Critic::Utils qw{ $SEVERITY_MEDIUM };

our $VERSION = '0.02';

Readonly::Scalar my $EXPL => q{Use PERL5LIB environment instead};

use constant supported_parameters => ();
use constant default_severity     => $SEVERITY_MEDIUM;
use constant default_themes       => qw(logiclab);
use constant applies_to           => 'PPI::Statement::Include';

sub violates {
    my ( $self, $elem ) = @_;

    my $child = $elem->schild(1);    #second token
    return if !$child;    #return if no token, this will not be relevant to us

    #second token should read: lib
    #See t/test.t for examples of variations
    $child =~ m{
        \A  #beginning of string
        lib #the word 'lib'
        \Z  #end of string
    }xsm or return;

    return $self->violation( q{Do not use 'use lib' statements}, $EXPL,
        $child );
}

1;

__END__

=pod

=head1 NAME

Perl::Critic::Policy::logicLAB::ProhibitUseLib - simple policy prohibiting the use of 'use lib'

=head1 AFFILIATION

This policy is a policy in the L<Perl::Critic::logicLAB> distribution.

=head1 VERSION

This documentation describes version 0.01

=head1 DESCRIPTION

The 'use lib' statement, hardcodes the include path to be used. This can give
issues when moving modules and scripts between diverse environments.

    use lib '/some/path';                                       #not ok
    use lib qw(/you/do/not/want/to/go/down/this/path /or/this); #not ok

Instead use the environment variable PERL5LIB

    #bash
    export PERL5LIB='/some/path/some/where'
    
    #tcsh and csh
    setenv PERL5LIB '/some/path/some/where'

=head1 CONFIGURATION AND ENVIRONMENT

This Policy is not configurable except for the standard options.
    
=head1 DEPENDENCIES AND REQUIREMENTS

=over

=item * L<Perl::Critic>

=item * L<Perl::Critic::Utils>

=item * L<Readonly>

=item * L<Test::More>

=item * L<Test::Perl::Critic>

=back

=head1 INCOMPATIBILITIES

This distribution has no known incompatibilities.

=head1 BUGS AND LIMITATIONS

Currently the policy has no special opinion on L<FindBin>. It only aims to
address messy, misleading, buggy and obscuring use of 'use lib'.

=head1 BUG REPORTING

Please use Requets Tracker for bug reporting:

    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-ProhibitUseLib

=head1 TEST AND QUALITY

The following policies have been disabled for this distribution

=over

=item * L<Perl::Crititc::Policy::ValuesAndExpressions::ProhibitConstantPragma>

=item * L<Perl::Crititc::Policy::NamingConventions::Capitalization>

=back

See also F<t/perlcriticrc>

=head2 TEST COVERAGE

    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    File                           stmt   bran   cond    sub    pod   time  total
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    ...ogicLAB/ProhibitUseLib.pm  100.0   50.0    n/a  100.0  100.0  100.0   95.3
    Total                         100.0   50.0    n/a  100.0  100.0  100.0   95.3
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    
=head1 SEE ALSO

=over

=item * L<http://perldoc.perl.org/perlrun.html#ENVIRONMENT>

=item * L<http://logiclab.jira.com/wiki/display/OPEN/Development#Development-MakeyourComponentsEnvironmentAgnostic>

=item * L<http://logicLAB.jira.com/wiki/display/PCPMPUL/Home>

=back

=head1 AUTHOR

=over

=item * Jonas B. Nielsen, jonasbn C<< <jonasbn@cpan.org> >>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2009-2011 Jonas B. Nielsen. All rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut
